### PR TITLE
Consumer redesign: Better backpressure, prefetching, reducer straggler mitigation.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Redesign's the shuffle consumer API, the consumer queue API and implementation, and the dataset's iterator to support better backpressure, shuffle output prefetching, and mitigation of reducer stragglers.

### CODE REVIEW

File: benchmarks/benchmark.py [hunk 1]
```diff
@@ -1,26 +1,96 @@
...
+            first_epoch = self._curr_epochs.popleft()
+            await self._epoch_done_evs[first_epoch].wait()
+        self._curr_epochs.append(epoch)
+        self._consume_timestamps[epoch][timeit.default_timer()] = 0
+        print(f"Starting epoch {epoch} on consumer {self._rank}.")
+
+    def consume(self, epoch, batch):
+        self._consume_timestamps[epoch][timeit.default_timer()] = len(batch)
+
+    def producer_done(self, epoch):
```
Comment (stephanie-wang) [inline benchmarks/benchmark.py:47]: Shouldn't we wait until the consumer is done, not the producer?

File: ray_shuffling_data_loader/batch_queue.py [hunk 2]
```diff
@@ -329,57 +352,128 @@ def connect_queue_actor(name, num_retries=5):
 
 
 class _QueueActor:
-    def __init__(self, num_queues, maxsize):
+    def __init__(
+            self, max_epochs, num_epochs, num_trainers, maxsize,
+            wait_for_trainers=True):
```
Comment (stephanie-wang) [inline ray_shuffling_data_loader/batch_queue.py:357]: I'm not sure if there's a situation where we'd want this to be False. How about just removing it and always just waiting for the trainers? Then we don't need the code to also wait for producers.

File: ray_shuffling_data_loader/batch_queue.py [hunk 3]
```diff
@@ -329,57 +352,128 @@ def connect_queue_actor(name, num_retries=5):
 
 
 class _QueueActor:
-    def __init__(self, num_queues, maxsize):
+    def __init__(
+            self, max_epochs, num_epochs, num_trainers, maxsize,
```
Comment (stephanie-wang) [inline ray_shuffling_data_loader/batch_queue.py:356]: Could we set maxsize automatically, maybe to the number of expected batches + number of consumers? I think it would be best if we could hide this from the user. I could see this potentially causing a deadlock if the user sets it too low.

File: ray_shuffling_data_loader/dataset.py [hunk 4]
```diff
@@ -116,12 +118,44 @@ def __iter__(self):
...
+                    # This can happen if the only item in the queue was the
+                    # done signal, None.
+                    continue
+
+            # This ray.wait() also serves as our prefetching method, where pull
+            # requests for all objects in pending will be made at this time.
+            ready, pending = ray.wait(
+                pending, num_returns=1, fetch_local=True)
+            # Fetch the underlying dataframe.
+            df = ray.get(ready[0])
```
Comment (stephanie-wang) [inline ray_shuffling_data_loader/dataset.py:155]: Is `ready` guaranteed to only have one ObjectRef?

File: ray_shuffling_data_loader/dataset.py [hunk 5]
```diff
@@ -116,12 +118,44 @@ def __iter__(self):
...
-            batches = self._batch_queue.get(queue_idx, block=True)
-            if batches is None:
-                break
-            df = ray.get(batches)
+            if not pending:
+                if is_done:
+                    # No more items in pending and is_done flag is set, so
+                    # there are no more queue items to consume.
+                    break
+                if outstanding_items > 0:
```
Comment (stephanie-wang) [inline ray_shuffling_data_loader/dataset.py:130]: I think it's a bit easier to understand if this goes at the end of the while loop body, after the processing is done. Then you also don't need to handle the last iteration differently.

**Review by stephanie-wang — COMMENTED** [review]
This makes a lot of sense! I like the new API for the consumer. Had some minor comments, mainly I'd like to know if we could remove some args from the top-level API.

File: benchmarks/benchmark.py (hunk 1 above)
Comment (clarkzinzow) [reply benchmarks/benchmark.py:47]: We need both in order to know when an epoch is truly done, i.e. for the backpressure semantics to be correct: we need to wait until the producer indicates that it will put no more batches into the queue, and then we need to wait until all items that are currently in the queue are done being processed by the consumers. If you only wait until the consumers are done, then the queue will consider the epoch to be finished when the queue becomes empty, so any batches the producer puts in the queue after that point will be lost. If you only wait until the producers are done, then the queue will consider the epoch to be finished when the producers are done putting batches into the queue, so any batches the consumers are still working on aren't considered to be part of the epoch, which would lead to suboptimal backpressure semantics.

The other option would be calculating the number of reducer outputs per epoch and determine whether the epoch is done via in-queue accounting, but I thought it was cleaner for the queue to expose APIs to the producer to indicate when there are no more batch puts, and the consumer to indicate when all batches are done being processed; that way, the queue doesn't need to know the number of reducer outputs and doesn't need to duplicate the shuffler's bookkeeping to determine when an epoch is "done".

File: ray_shuffling_data_loader/dataset.py (hunk 4 above)
Comment (clarkzinzow) [reply ray_shuffling_data_loader/dataset.py:155]: Yep, note the `num_returns=1` argument in the above `ray.wait()`.

File: ray_shuffling_data_loader/batch_queue.py (hunk 3 above)
Comment (clarkzinzow) [reply ray_shuffling_data_loader/batch_queue.py:356]: I actually don't think that we should even set `maxsize`, since there doesn't seem to be an advantage to setting any queue size, right? Our backpressure is currently designed to happen pre-put, where we push back on starting a new epoch shuffle, not on putting batches into the queue.

I'm definitely happy to remove this as a configurable parameter, I agree that it shouldn't be exposed to the end-user.

File: ray_shuffling_data_loader/batch_queue.py (hunk 2 above)
Comment (clarkzinzow) [reply ray_shuffling_data_loader/batch_queue.py:357]: >I'm not sure if there's a situation where we'd want this to be False. How about just removing it and always just waiting for the trainers?

Ah, yeah, this was a hold-over from when the benchmarking script used this `BatchQueue` as a sink and didn't want to wait for the consumers of the queue. I'll remove this!

>Then we don't need the code to also wait for producers.

As I wrote above, we'd still need to wait for the producers to know when no more items will be added to the queue.

File: ray_shuffling_data_loader/dataset.py (hunk 5 above)
Comment (clarkzinzow) [reply ray_shuffling_data_loader/dataset.py:130]: Agreed that it might be easier to understand, although putting it at the beginning ended up consolidating a lot of logic. The current (optimized with batching) flow is:
1. Get a batch of items from the queue.
2. If next item is not the done sentinel, get the next item. Otherwise, break and process leftover item.
3. Process item.
4. If items left in current batch, go to 2. Otherwise, signal to queue that current batch is done being processed, and go to 1.

Maybe this could be restructured with nested while loops...
```python
while True:
    while pending:
        // Get and process items.
    // Signal to queue that current batch is done being processed.
    // If done flag is set, break and process leftover item.
    // Get batch of items from queue.
```

So that would move the getting/processing of items to the start of the while loop body, and should eliminate the `continue` and the `task_done` outside of the outer while loop.

Comment (clarkzinzow) [reply ray_shuffling_data_loader/dataset.py]: I ended up opting for
```python
while not is_done:
    // Get batch of items from queue.
    while pending:
        // Get and process items.
    // Signal to queue that current batch is done being processed.
// Process leftover item.

File: benchmarks/benchmark.py (hunk 1 above)
Comment (stephanie-wang) [reply benchmarks/benchmark.py:58]: Thanks for the explanation, that makes sense! Could you add a comment to this effect in the queue actor's `new_epoch` implementation and the class interface (that callers should first wait for producers then consumers to finish)?

**Comment by stephanie-wang** [comment]
Thanks for the fixes! Just asked for comments, otherwise feel free to merge whenever.

File: benchmarks/benchmark.py (hunk 1 above)
Comment (clarkzinzow) [reply benchmarks/benchmark.py:58]: Sure, can do!

Comment (clarkzinzow) [reply benchmarks/benchmark.py:52]: Btw, the waiting for the producers and then waiting for the consumers is an implementation detail of `new_epoch`, so I'm just going to indicate that the data loaders should be sure to use `producer_done()` to indicate that the batch producer is done, and that the trainers should be sure to use `task_done()` to indicate that the batch consumers are done.
